File: q1043/q1043_recursive.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        memo = [-1 for _ in range(len(arr))]
        return self.dpHelper(arr, k, 0, memo)
        
    # let dpHelper return the max after partitioning if we start from index start
    def dpHelper(self, arr, k, start, memo): 
        # base case: return 0 if start goes out of the boundary
        if start >= len(arr): return 0

        # memoized case: if we have already computed this solution
        if memo[start] != -1: return memo[start]

        # recursive case: initialize the current subarray max as arr[start]
        currMax = arr[start]
        res = 0
        for end in range(start, min(len(arr), start + k)):
            # Update the current subarray max
            currMax = max(currMax, arr[end])

            # compute the current subarry sum
            subarraySum = (end - start + 1) * currSubarrayMax

            # recursively compute the result if start from index end+1
            res = max(res, subarraySum + self.dpHelper(arr, k, end + 1, memo))
        
        memo[start] = res
        return res

File: q1043/q1043_recursive.java
class Solution {
    public int maxSumAfterPartitioning(int[] arr, int k) {
        int[] memo = new int[arr.length];
        Arrays.fill(memo, -1);
        return dpHelper(arr, k, 0, memo);
    }

    // let dpHelper return the max after partitioning if we start from index start
    private int dpHelper(int[] arr, int k, int start, int[] memo) {
        // base case: return 0 if start goes out of the boundary
        if (start >= arr.length)
            return 0;

        // memoized case: if we have already computed this solution
        if (memo[start] != -1)
            return memo[start];

        // recursive case: initialize the current subarray max as arr[start]
        int currMax = arr[start];
        for (int end = start; end < Math.min(arr.length, start + k); end++) {
            // Update the current subarray max
            currMax = Math.max(currMax, arr[end]);

            // compute the current subarry sum
            int subarraySum = (end - start + 1) * currMax;

            // recursively compute the result if start from index end+1
            memo[start] = Math.max(memo[start], subarraySum + dpHelper(arr, k, end + 1, memo));
        }

        return memo[start];
    }
}